File: python/aethernet/transport/lora.py
# SPDX-License-Identifier: MIT

"""
Real LoRa (Aether Red / CircleLink) transport over a serial-attached RYLR-class module.

Speaks the RYLR-class AT command set (Reyax RYLR896/RYLR998 and compatibles) on an
SX127x/SX126x radio, mirroring the C#/Go/Rust/C `LoRaSerialTransport`: opens the serial
port, configures the radio, sends with ``AT+SEND``, and surfaces inbound ``+RCV`` frames.

Requires the optional ``pyserial`` dependency::

    pip install "aethernet-protocol[lora]"   # or: pip install pyserial

Verification status: real driver; imports lazily so the core package still imports without
pyserial. Runtime-UNVERIFIED — not exercised against a physical module.
"""

import asyncio
import threading
from dataclasses import dataclass
from typing import Callable, Dict, List, Optional

from aethernet.transport.per_transport_metrics import PerTransportMetrics
from aethernet.transport.transport_service import TransportService


@dataclass
class LoRaOptions:
    """Configuration for a RYLR-class serial LoRa module."""

    port_name: str  # "COM5" or "/dev/ttyUSB0" — required
    baud_rate: int = 115200
    address: int = 1  # this node's LoRa address (1-65535)
    network_id: int = 18  # RYLR network id
    band_hz: int = 868_500_000  # EU868; US915 = 915_000_000
    spreading_factor: int = 9  # 7-12
    bandwidth_index: int = 7  # 7=125kHz, 8=250, 9=500
    coding_rate: int = 1  # 1=4/5
    preamble_length: int = 12


class LoRaSerialTransport(TransportService):
    """LoRa transport over a serial-attached RYLR-class module."""

    def __init__(self, options: LoRaOptions) -> None:
        self._opts = options
        self._metrics = PerTransportMetrics()
        self._lock = threading.Lock()
        self._serial: Optional[object] = None
        self._available = False
        self._running = False
        self._callbacks: List[Callable[[str, bytes], None]] = []
        self._peer_addrs: Dict[str, int] = {}
        self._reader: Optional[threading.Thread] = None

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    def open(self) -> None:
        """Open the serial port, configure the radio, and start the reader thread."""
        try:
            import serial  # pyserial — optional dependency
        except ImportError as exc:  # pragma: no cover - import-guard
            raise RuntimeError(
                "LoRa transport requires pyserial: pip install \"aethernet-protocol[lora]\""
            ) from exc

        with self._lock:
            if self._available:
                return
            self._serial = serial.Serial(self._opts.port_name, self._opts.baud_rate, timeout=1)
            self._configure()
            self._running = True
            self._available = True
            self._reader = threading.Thread(target=self._read_loop, name="lora-reader", daemon=True)
            self._reader.start()

    def _configure(self) -> None:
        assert self._serial is not None
        for cmd in (
            f"AT+ADDRESS={self._opts.address}",
            f"AT+NETWORKID={self._opts.network_id}",
            f"AT+BAND={self._opts.band_hz}",
            f"AT+PARAMETER={self._opts.spreading_factor},{self._opts.bandwidth_index},"
            f"{self._opts.coding_rate},{self._opts.preamble_length}",
        ):
            self._serial.write((cmd + "\r\n").encode("ascii"))  # type: ignore[attr-defined]

    def close(self) -> None:
        """Stop the reader thread and close the serial port."""
        with self._lock:
            self._running = False
            self._available = False
            ser = self._serial
            self._serial = None
        if ser is not None:
            try:
                ser.close()  # type: ignore[attr-defined]
            except Exception:
                pass

    # ── TransportService ───────────────────────────────────────────────────────

    @property
    def name(self) -> str:
        return "Aether Red (LoRa/CircleLink)"

    @property
    def is_available(self) -> bool:
        with self._lock:
            return self._available

    @property
    def max_bandwidth_bps(self) -> int:
        return 37_500  # SF7/BW125 ~= 37.5 kbps

    @property
    def max_range_meters(self) -> int:
        return 15_000  # up to ~15 km LOS

    @property
    def power_cost_relative(self) -> int:
        return 8  # high TX power (1-10 scale)

    @property
    def max_concurrent_peers(self) -> int:
        return 255

    @property
    def metrics(self) -> PerTransportMetrics:
        return self._metrics

    async def send_async(self, peer_uhid: str, data: bytes) -> bool:
        if not data:
            return False
        with self._lock:
            if not self._available or self._serial is None:
                return False
            addr = self._peer_addrs.get(peer_uhid, 0)  # 0 = broadcast (managed-flood mesh)
            ser = self._serial
        # Hex-encode so the payload survives the AT text protocol; length field is the hex length.
        payload = data.hex().upper()
        cmd = f"AT+SEND={addr},{len(payload)},{payload}\r\n"
        try:
            ser.write(cmd.encode("ascii"))  # type: ignore[attr-defined]
            self._metrics.record_sample(0, True, len(data))
            return True
        except Exception:
            self._metrics.record_sample(0, False, 0)
            return False

    async def send_stream_async(self, peer_uhid: str, data_stream: asyncio.StreamReader) -> bool:
        data = await data_stream.read()
        return await self.send_async(peer_uhid, data)

    def is_connected(self, peer_uhid: str) -> bool:
        return self.is_available  # connectionless broadcast medium

    def on_data_received(self, callback: Callable[[str, bytes], None]) -> None:
        with self._lock:
            self._callbacks.append(callback)

    def register_peer(self, peer_uhid: str, address: int) -> None:
        """Map an AetherNet peer UHID to a numeric LoRa node address (1-65535) for directed sends."""
        if not peer_uhid:
            return
        with self._lock:
            self._peer_addrs[peer_uhid] = address

    # ── Receive ────────────────────────────────────────────────────────────────

    def _read_loop(self) -> None:
        while True:
            with self._lock:
                running = self._running
                ser = self._serial
            if not running or ser is None:
                break
            try:
                line = ser.readline()  # type: ignore[attr-defined]
            except Exception:
                break
            if not line:
                continue
            self._handle_line(line.decode("ascii", errors="ignore").strip())

    def _handle_line(self, line: str) -> None:
        # RYLR inbound frame: +RCV=<address>,<length>,<hexdata>,<rssi>,<snr>
        if not line.startswith("+RCV="):
            return
        parts = line[5:].split(",")
        if len(parts) < 3:
            return
        try:
            addr = int(parts[0])
            data = bytes.fromhex(parts[2])
        except ValueError:
            return
        with self._lock:
            callbacks = list(self._callbacks)
        for callback in callbacks:
            callback(str(addr), data)

File: go/transport/lora/lora.go
// SPDX-License-Identifier: MIT

// Package lora is a real LoRa (Aether Red / CircleLink) transport over a serial-attached LoRa module
// that speaks the RYLR-class AT command set (Reyax RYLR896/RYLR998 and compatibles) on an SX127x/
// SX126x radio. It mirrors the C# LoRaSerialTransportService: it opens the serial port, configures the
// radio, sends with AT+SEND, and surfaces inbound +RCV frames.
//
// Verification status: this is a real driver and it compiles, but it is runtime-UNVERIFIED — it has
// not been exercised against a physical module. On-radio bring-up (two modules exchanging a frame) is
// the open step. IsAvailable reflects whether the configured serial port actually opened.
package lora

import (
	"bufio"
	"context"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"sync"

	"go.bug.st/serial"

	"github.com/bhengubv/aether-protocol/go/transport"
)

// Options configures a RYLR-class serial LoRa module.
type Options struct {
	PortName        string // "COM5" or "/dev/ttyUSB0"
	BaudRate        int    // default 115200
	Address         uint16 // this node's LoRa address (1-65535)
	NetworkID       int    // RYLR network id
	BandHz          int64  // EU868 = 868500000; US915 = 915000000
	SpreadingFactor int    // 7-12
	BandwidthIndex  int    // 7=125kHz, 8=250, 9=500
	CodingRate      int    // 1=4/5
	PreambleLength  int
}

func (o *Options) withDefaults() {
	if o.BaudRate == 0 {
		o.BaudRate = 115200
	}
	if o.Address == 0 {
		o.Address = 1
	}
	if o.NetworkID == 0 {
		o.NetworkID = 18
	}
	if o.BandHz == 0 {
		o.BandHz = 868_500_000
	}
	if o.SpreadingFactor == 0 {
		o.SpreadingFactor = 9
	}
	if o.BandwidthIndex == 0 {
		o.BandwidthIndex = 7
	}
	if o.CodingRate == 0 {
		o.CodingRate = 1
	}
	if o.PreambleLength == 0 {
		o.PreambleLength = 12
	}
}

// LoRaSerialTransport implements transport.TransportService over a serial LoRa module.
type LoRaSerialTransport struct {
	opts    Options
	metrics *transport.PerTransportMetrics

	mu        sync.Mutex
	port      serial.Port
	available bool
	onData    func(peerUhid string, data []byte)
	peerAddrs map[string]uint16
	stop      chan struct{}
}

var _ transport.TransportService = (*LoRaSerialTransport)(nil)

// New creates a transport for the given module options (call Open to bring it up).
func New(opts Options) *LoRaSerialTransport {
	opts.withDefaults()
	return &LoRaSerialTransport{
		opts:      opts,
		metrics:   transport.NewPerTransportMetrics(),
		peerAddrs: make(map[string]uint16),
	}
}

// OnDataReceived registers the handler for inbound bytes (the receive surface).
func (t *LoRaSerialTransport) OnDataReceived(h func(peerUhid string, data []byte)) {
	t.mu.Lock()
	t.onData = h
	t.mu.Unlock()
}

// Open opens the serial port and configures the radio. Sets IsAvailable on success.
func (t *LoRaSerialTransport) Open() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.available {
		return nil
	}
	port, err := serial.Open(t.opts.PortName, &serial.Mode{BaudRate: t.opts.BaudRate})
	if err != nil {
		return fmt.Errorf("lora: open %s: %w", t.opts.PortName, err)
	}
	t.port = port
	t.configure(port)
	t.stop = make(chan struct{})
	t.available = true
	go t.readLoop(port, t.stop)
	return nil
}

func (t *LoRaSerialTransport) configure(port serial.Port) {
	cmds := []string{
		fmt.Sprintf("AT+ADDRESS=%d", t.opts.Address),
		fmt.Sprintf("AT+NETWORKID=%d", t.opts.NetworkID),
		fmt.Sprintf("AT+BAND=%d", t.opts.BandHz),
		fmt.Sprintf("AT+PARAMETER=%d,%d,%d,%d",
			t.opts.SpreadingFactor, t.opts.BandwidthIndex, t.opts.CodingRate, t.opts.PreambleLength),
	}
	for _, cmd := range cmds {
		_, _ = port.Write([]byte(cmd + "\r\n"))
	}
}

// ── transport.TransportService ───────────────────────────────────────────────

func (t *LoRaSerialTransport) Name() string { return "Aether Red (LoRa/CircleLink)" }

func (t *LoRaSerialTransport) IsAvailable() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.available
}

func (t *LoRaSerialTransport) MaxBandwidthBps() int64    { return 37_500 } // SF7/BW125 ≈ 37.5 kbps
func (t *LoRaSerialTransport) MaxRangeMeters() int32     { return 15_000 } // up to ~15 km LOS
func (t *LoRaSerialTransport) PowerCostRelative() int32  { return 8 }      // high TX power (1-10 scale)
func (t *LoRaSerialTransport) MaxConcurrentPeers() int32 { return 255 }

func (t *LoRaSerialTransport) Metrics() *transport.PerTransportMetrics { return t.metrics }

func (t *LoRaSerialTransport) SendAsync(_ context.Context, peerUhid string, data []byte) (bool, error) {
	if len(data) == 0 {
		return false, fmt.Errorf("lora: data cannot be empty")
	}
	t.mu.Lock()
	if !t.available || t.port == nil {
		t.mu.Unlock()
		return false, fmt.Errorf("lora: transport not available")
	}
	addr := uint16(0) // 0 = broadcast (managed-flood mesh)
	if a, ok := t.peerAddrs[peerUhid]; ok {
		addr = a
	}
	port := t.port
	t.mu.Unlock()

	payload := strings.ToUpper(hex.EncodeToString(data)) // hex so it survives the AT text protocol
	cmd := fmt.Sprintf("AT+SEND=%d,%d,%s\r\n", addr, len(payload), payload)
	if _, err := port.Write([]byte(cmd)); err != nil {
		t.metrics.RecordSample(0, false, 0)
		return false, err
	}
	t.metrics.RecordSample(0, true, int64(len(data)))
	return true, nil
}

func (t *LoRaSerialTransport) SendStreamAsync(ctx context.Context, peerUhid string, data []byte) (bool, error) {
	return t.SendAsync(ctx, peerUhid, data)
}

func (t *LoRaSerialTransport) IsConnected(string) bool { return t.IsAvailable() } // connectionless

// RegisterPeer maps an AetherNet peer UHID to a numeric LoRa node address (1-65535) for directed sends.
func (t *LoRaSerialTransport) RegisterPeer(peerUhid string, address uint16) {
	if peerUhid == "" {
		return
	}
	t.mu.Lock()
	t.peerAddrs[peerUhid] = address
	t.mu.Unlock()
}

// Close stops the read loop and closes the serial port.
func (t *LoRaSerialTransport) Close() error {
	t.mu.Lock()
	if !t.available {
		t.mu.Unlock()
		return nil
	}
	t.available = false
	if t.stop != nil {
		close(t.stop)
		t.stop = nil
	}
	port := t.port
	t.port = nil
	t.mu.Unlock()
	if port != nil {
		return port.Close()
	}
	return nil
}

func (t *LoRaSerialTransport) readLoop(port serial.Port, stop chan struct{}) {
	scanner := bufio.NewScanner(port)
	for scanner.Scan() {
		select {
		case <-stop:
			return
		default:
		}
		t.handleLine(strings.TrimSpace(scanner.Text()))
	}
}

func (t *LoRaSerialTransport) handleLine(line string) {
	// RYLR inbound frame: +RCV=<address>,<length>,<hexdata>,<rssi>,<snr>
	if !strings.HasPrefix(line, "+RCV=") {
		return
	}
	parts := strings.Split(line[5:], ",")
	if len(parts) < 3 {
		return
	}
	addr, err := strconv.ParseUint(parts[0], 10, 16)
	if err != nil {
		return
	}
	data, err := hex.DecodeString(parts[2])
	if err != nil {
		return
	}
	t.mu.Lock()
	h := t.onData
	t.mu.Unlock()
	if h != nil {
		h(strconv.FormatUint(addr, 10), data)
	}
}
